x/erc20/keeper/denom_metadata_hooks.go

```diff
@@ -0,0 +1,39 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+type ERC20BankContractRegistrationHook struct {
+	erc20Keeper Keeper
+}
+
+// NewERC20ContractRegistrationHook returns the DenomMetadataHooks for ERC20 registration
+func NewERC20ContractRegistrationHook(erc20Keeper Keeper) ERC20BankContractRegistrationHook {
+	return ERC20BankContractRegistrationHook{
+		erc20Keeper: erc20Keeper,
+	}
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataCreation(ctx sdk.Context, newDenomMetadata banktypes.Metadata) error {
+	// only deploy for IBC denom.
+	if !strings.HasPrefix(strings.ToLower(newDenomMetadata.Base), "ibc/") {
+		return nil
+	}
+
+	if _, err := e.erc20Keeper.RegisterCoin(ctx, newDenomMetadata); err != nil {
+		return fmt.Errorf("failed to deploy the erc20 contract for the IBC coin: %s; error: %w", newDenomMetadata.Base, err)
+	}
+
+	return nil
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataUpdate(sdk.Context, banktypes.Metadata) error {
+	// do nothing
```

Comment:
is it really fine to do nothing? on the face of it, we'd need to update the contract or something

Reply:
Can we do that? Smart contracts are immutable by design

Reply:
I suggest returning an error here if trying to update a metadata with an already existing contract.
It should fail the update tx as well.
I think it's better than the alternative, where we have difference between the bank module and the erc20 module